File: clients/python3/dice52/handshake.py
"""Handshake protocol using hybrid KEM (Kyber768 + X25519) and Dilithium3."""

from typing import Tuple

from kyber_py.kyber import Kyber768
from dilithium_py.dilithium import Dilithium3
from cryptography.hazmat.primitives.asymmetric.x25519 import X25519PrivateKey, X25519PublicKey

from .types import HandshakeMessage, RatchetMessage, SIG_CONTEXT
from .kdf import derive_hybrid_shared_secret
from .error import InvalidHandshakeSignature, KemError


def generate_x25519_keypair() -> Tuple[bytes, bytes]:
    """
    Generate a new X25519 key pair.
    
    Returns:
        Tuple of (public_key, private_key)
    """
    private_key = X25519PrivateKey.generate()
    public_key = private_key.public_key()
    return (
        public_key.public_bytes_raw(),
        private_key.private_bytes_raw()
    )


def x25519_shared_secret(our_priv: bytes, peer_pub: bytes) -> bytes:
    """
    Compute X25519 shared secret.
    
    Args:
        our_priv: Our X25519 private key (32 bytes)
        peer_pub: Peer's X25519 public key (32 bytes)
        
    Returns:
        Shared secret (32 bytes)
    """
    private_key = X25519PrivateKey.from_private_bytes(our_priv)
    public_key = X25519PublicKey.from_public_bytes(peer_pub)
    return private_key.exchange(public_key)


def generate_kem_keypair() -> Tuple[bytes, bytes]:
    """
    Generate a new Kyber768 key pair.
    
    Returns:
        Tuple of (public_key, private_key)
    """
    public_key, private_key = Kyber768.keygen()
    return public_key, private_key


def generate_signing_keypair() -> Tuple[bytes, bytes]:
    """
    Generate a new Dilithium3 key pair.
    
    Returns:
        Tuple of (public_key, private_key)
    """
    public_key, private_key = Dilithium3.keygen()
    return public_key, private_key


def initiator_encapsulate(peer_pub: bytes) -> Tuple[bytes, bytes]:
    """
    Initiator encapsulate: Alice encapsulates to Bob's public key (Section 7).
    
    Args:
        peer_pub: Peer's Kyber public key
        
    Returns:
        Tuple of (shared_secret, ciphertext)
    """
    shared_secret, ciphertext = Kyber768.encaps(peer_pub)
    return shared_secret, ciphertext


def responder_decapsulate(our_priv: bytes, ct: bytes) -> bytes:
    """
    Responder decapsulate: Bob decapsulates using his private key (Section 7).
    
    Args:
        our_priv: Our Kyber private key
        ct: Received ciphertext
        
    Returns:
        Shared secret
    """
    shared_secret = Kyber768.decaps(our_priv, ct)
    return shared_secret


def initiator_handshake(
    peer_kem: bytes,
    peer_ecdh: bytes,
    id_priv: bytes,
) -> Tuple[HandshakeMessage, bytes, bytes]:
    """
    Initiator handshake: hybrid encapsulation to peer's public keys and sign (Section 6.2).
    
    Args:
        peer_kem: Peer's Kyber public key
        peer_ecdh: Peer's X25519 public key
        id_priv: Our Dilithium private key for signing
        
    Returns:
        Tuple of (handshake_message, hybrid_shared_secret, our_ecdh_private_key)
    """
    # Generate ephemeral X25519 key pair
    ecdh_pub, ecdh_priv = generate_x25519_keypair()
    
    # Kyber encapsulation
    ss_pq, ct = initiator_encapsulate(peer_kem)
    
    # X25519 key agreement
    ss_ecdh = x25519_shared_secret(ecdh_priv, peer_ecdh)
    
    # Derive hybrid shared secret
    ss_hybrid = derive_hybrid_shared_secret(ss_pq, ss_ecdh)
    
    # Sign: context || ciphertext || ecdh_pub
    to_sign = SIG_CONTEXT + ct + ecdh_pub
    
    signature = Dilithium3.sign(id_priv, to_sign)
    
    return HandshakeMessage(kyber_ct=ct, ecdh_pub=ecdh_pub, sig=signature), ss_hybrid, ecdh_priv


def responder_handshake(
    msg: HandshakeMessage,
    kem_priv: bytes,
    ecdh_priv: bytes,
    peer_id: bytes,
) -> bytes:
    """
    Responder handshake: verify signature and hybrid decapsulation (Section 6.3).
    
    Args:
        msg: Received handshake message
        kem_priv: Our Kyber private key
        ecdh_priv: Our X25519 private key
        peer_id: Peer's Dilithium public key for verification
        
    Returns:
        Hybrid shared secret
        
    Raises:
        InvalidHandshakeSignature: If signature verification fails
    """
    # Verify signature over context || ciphertext || ecdh_pub
    to_verify = SIG_CONTEXT + msg.kyber_ct + msg.ecdh_pub
    
    if not Dilithium3.verify(peer_id, to_verify, msg.sig):
        raise InvalidHandshakeSignature("Handshake signature verification failed")
    
    # Decapsulate Kyber
    ss_pq = responder_decapsulate(kem_priv, msg.kyber_ct)
    
    # X25519 key agreement
    ss_ecdh = x25519_shared_secret(ecdh_priv, msg.ecdh_pub)
    
    # Derive hybrid shared secret
    ss_hybrid = derive_hybrid_shared_secret(ss_pq, ss_ecdh)
    
    return ss_hybrid


def initiator_hybrid_encapsulate(peer_kem: bytes, peer_ecdh: bytes) -> Tuple[bytes, bytes, bytes, bytes]:
    """
    Initiator hybrid encapsulate: Alice encapsulates to Bob's public keys.
    
    Args:
        peer_kem: Peer's Kyber public key
        peer_ecdh: Peer's X25519 public key
        
    Returns:
        Tuple of (hybrid_shared_secret, kyber_ciphertext, our_ecdh_pub, our_ecdh_priv)
    """
    # Generate ephemeral X25519 key pair
    ecdh_pub, ecdh_priv = generate_x25519_keypair()
    
    # Kyber encapsulation
    ss_pq, ct = initiator_encapsulate(peer_kem)
    
    # X25519 key agreement
    ss_ecdh = x25519_shared_secret(ecdh_priv, peer_ecdh)
    
    # Derive hybrid shared secret
    ss_hybrid = derive_hybrid_shared_secret(ss_pq, ss_ecdh)
    
    return ss_hybrid, ct, ecdh_pub, ecdh_priv


def responder_hybrid_decapsulate(
    kem_priv: bytes,
    ecdh_priv: bytes,
    kyber_ct: bytes,
    peer_ecdh_pub: bytes
) -> bytes:
    """
    Responder hybrid decapsulate: Bob decapsulates using his private keys.
    
    Args:
        kem_priv: Our Kyber private key
        ecdh_priv: Our X25519 private key
        kyber_ct: Received Kyber ciphertext
        peer_ecdh_pub: Peer's X25519 public key
        
    Returns:
        Hybrid shared secret
    """
    # Decapsulate Kyber
    ss_pq = responder_decapsulate(kem_priv, kyber_ct)
    
    # X25519 key agreement
    ss_ecdh = x25519_shared_secret(ecdh_priv, peer_ecdh_pub)
    
    # Derive hybrid shared secret
    ss_hybrid = derive_hybrid_shared_secret(ss_pq, ss_ecdh)
    
    return ss_hybrid

File: clients/golang/pkg/dice52/handshake.go
package dice52

import (
	"crypto/rand"
	"errors"

	"github.com/cloudflare/circl/dh/x25519"
	"github.com/cloudflare/circl/kem/kyber/kyber768"
	"github.com/cloudflare/circl/sign/dilithium/mode3"
)

// GenerateX25519Keypair generates a new X25519 key pair
func GenerateX25519Keypair() (pub, priv []byte, err error) {
	var privKey x25519.Key
	var pubKey x25519.Key

	if _, err := rand.Read(privKey[:]); err != nil {
		return nil, nil, err
	}

	x25519.KeyGen(&pubKey, &privKey)

	return pubKey[:], privKey[:], nil
}

// X25519SharedSecret computes the X25519 shared secret
func X25519SharedSecret(ourPriv, peerPub []byte) ([]byte, error) {
	if len(ourPriv) != 32 || len(peerPub) != 32 {
		return nil, errors.New("invalid X25519 key length")
	}

	var priv, pub, ss x25519.Key
	copy(priv[:], ourPriv)
	copy(pub[:], peerPub)

	if !x25519.Shared(&ss, &priv, &pub) {
		return nil, errors.New("X25519 shared secret computation failed")
	}

	return ss[:], nil
}

// InitiatorHandshake is used to send the handshake message to the peer using hybrid KEM (Section 6.2)
// Returns the handshake message, hybrid shared secret, and our X25519 private key for session storage
func InitiatorHandshake(peerKEM *kyber768.PublicKey, peerECDH []byte, idPriv *mode3.PrivateKey) (*HandshakeMessage, []byte, []byte, error) {
	// Step 1: Generate ephemeral X25519 key pair
	ecdhPub, ecdhPriv, err := GenerateX25519Keypair()
	if err != nil {
		return nil, nil, nil, err
	}

	// Step 2: Kyber encapsulation
	ct, ssPQ, err := kyber768.Scheme().Encapsulate(peerKEM)
	if err != nil {
		return nil, nil, nil, err
	}

	// Step 3: X25519 key agreement
	ssECDH, err := X25519SharedSecret(ecdhPriv, peerECDH)
	if err != nil {
		return nil, nil, nil, err
	}

	// Step 4: Derive hybrid shared secret
	ssHybrid := DeriveHybridSharedSecret(ssPQ, ssECDH)

	// Step 5: Sign all handshake inputs (SigContext || KyberCT || ECDHPub)
	toSign := append([]byte(SigContext), ct...)
	toSign = append(toSign, ecdhPub...)
	sig := make([]byte, mode3.SignatureSize)
	mode3.SignTo(idPriv, toSign, sig)

	return &HandshakeMessage{KyberCT: ct, ECDHPub: ecdhPub, Sig: sig}, ssHybrid, ecdhPriv, nil
}

// ResponderHandshake is used to receive the handshake message from the peer using hybrid KEM (Section 6.3)
// Returns the hybrid shared secret
func ResponderHandshake(msg *HandshakeMessage, kemPriv *kyber768.PrivateKey, ecdhPriv []byte, peerID *mode3.PublicKey) ([]byte, error) {
	// Step 1: Verify signature over SigContext || KyberCT || ECDHPub
	toVerify := append([]byte(SigContext), msg.KyberCT...)
	toVerify = append(toVerify, msg.ECDHPub...)
	if !mode3.Verify(peerID, toVerify, msg.Sig) {
		return nil, errors.New("handshake signature invalid")
	}

	// Step 2: Decapsulate Kyber ciphertext
	ssPQ, err := kyber768.Scheme().Decapsulate(kemPriv, msg.KyberCT)
	if err != nil {
		return nil, err
	}

	// Step 3: X25519 key agreement
	ssECDH, err := X25519SharedSecret(ecdhPriv, msg.ECDHPub)
	if err != nil {
		return nil, err
	}

	// Step 4: Derive hybrid shared secret
	ssHybrid := DeriveHybridSharedSecret(ssPQ, ssECDH)

	return ssHybrid, nil
}

// HybridEncapsulateResult holds the result of hybrid encapsulation
type HybridEncapsulateResult struct {
	SSHybrid []byte // Combined hybrid shared secret
	KyberCT  []byte // Kyber ciphertext
	ECDHPub  []byte // Our X25519 public key
	ECDHPriv []byte // Our X25519 private key (for session storage)
}

// InitiatorHybridEncapsulate: Alice encapsulates to Bob's public keys using hybrid KEM
// Returns the hybrid shared secret, Kyber CT, and X25519 public key
func InitiatorHybridEncapsulate(peerKEM *kyber768.PublicKey, peerECDH []byte) (*HybridEncapsulateResult, error) {
	// Generate ephemeral X25519 key pair
	ecdhPub, ecdhPriv, err := GenerateX25519Keypair()
	if err != nil {
		return nil, err
	}

	// Kyber encapsulation
	ct, ssPQ, err := kyber768.Scheme().Encapsulate(peerKEM)
	if err != nil {
		return nil, err
	}

	// X25519 key agreement
	ssECDH, err := X25519SharedSecret(ecdhPriv, peerECDH)
	if err != nil {
		return nil, err
	}

	// Derive hybrid shared secret
	ssHybrid := DeriveHybridSharedSecret(ssPQ, ssECDH)

	return &HybridEncapsulateResult{
		SSHybrid: ssHybrid,
		KyberCT:  ct,
		ECDHPub:  ecdhPub,
		ECDHPriv: ecdhPriv,
	}, nil
}

// ResponderHybridDecapsulate: Bob decapsulates using his private keys (hybrid KEM)
// Returns the hybrid shared secret
func ResponderHybridDecapsulate(kemPriv *kyber768.PrivateKey, ecdhPriv []byte, kyberCT, peerECDHPub []byte) ([]byte, error) {
	// Decapsulate Kyber ciphertext
	ssPQ, err := kyber768.Scheme().Decapsulate(kemPriv, kyberCT)
	if err != nil {
		return nil, err
	}

	// X25519 key agreement
	ssECDH, err := X25519SharedSecret(ecdhPriv, peerECDHPub)
	if err != nil {
		return nil, err
	}

	// Derive hybrid shared secret
	ssHybrid := DeriveHybridSharedSecret(ssPQ, ssECDH)

	return ssHybrid, nil
}

// Legacy functions for backwards compatibility (Kyber-only)

// InitiatorEncapsulate: Alice encapsulates to Bob's public key (Kyber-only, legacy)
// Returns the shared secret and ciphertext to send to Bob
func InitiatorEncapsulate(peerPub *kyber768.PublicKey) (ss, ct []byte, err error) {
	ct, ss, err = kyber768.Scheme().Encapsulate(peerPub)
	return ss, ct, err
}

// ResponderDecapsulate: Bob decapsulates using his private key (Kyber-only, legacy)
func ResponderDecapsulate(ourPriv *kyber768.PrivateKey, ct []byte) ([]byte, error) {
	return kyber768.Scheme().Decapsulate(ourPriv, ct)
}
